### DISCUSSION TITLE
DCOM authentication level enforcement

### DISCUSSION THREAD

**Discussion by ldariva** [iterativ/openopc2#47] [General]
I notice that you use the dll from gray-box (http://gray-box.net/). This dll is not compatible with the nem DCOM patch that force the minimun level of authentication. Is there any other dll that is compatible with this new requirements? I believe that gray-box won't release a new version of this dll as it is a russian based company and may be suffering with the current situation on Europe.

https://support.microsoft.com/en-us/topic/kb5004442-manage-changes-for-windows-dcom-server-security-feature-bypass-cve-2021-26414-f1400b52-c141-43d2-941e-37ed901c769c

**Comment by renzop** [comment]
Good, point, but  I would drop DCOM support anyway. This is a legacy feature from the original OpenOPC. We have never tested it, and what I read about it is that it is quite difficult to configure it. 

So the combination of using COM and the OpenOPC 2 gateway you can achieve a similar benefit for the user.  

Also I would like to reduce the complexity of OpenOPC2 as much as possible, since there is too much stuff that was added in the original library. Which is hard to test and maintain.

**Reply by ldariva** [reply]
The problem is that many systems still rely on Classic OPC, which means, DCOM. And the gateway has a problem which, despite it works (and I have already tested it) many suppliers won't support you if you install any software not homologated by them and a open source ones are particular difficult to homologate. One thing is using it on a client application, another is installling it on a server.

Do you know if there is any other dll that could be used to replace his gbda_aut.dll?

**Reply by renzop** [reply]
We have to distinguish two things here, OpenOPC2 is always Opc Classic (OPC UA would be a different story). 

The question of COM and DCOM or COM only is the topic here. I see your point in homologation. But for the whole DCOM part we have no tests yet. Would you like to participate and work on that? 

I don't know if there is an alternative to the dlls from graybox, this is an area where we did not invest any time  in yet. Let me know if you find anything. Maybe ask graybox directly. I think getting patches and other updates in general is a problem with Opc Classic stuff since it is all very old and it comes from a time were IT Security was something from the movies.

**Reply by ldariva** [reply]
Hi. I tried to send a e-mail to them but on doing that I realized that they are a russian company. I didn't received any response from them and I suspect that this is due to the currenct geopolitical situation over there. 
I would love to participate on trying to find a solution for that. I am suprised that no one else is talking about this issue. 
I can test any proposed solution as I have access to many industrial opc servers here in my company.
I found this link about OPC with Go (https://github.com/konimarti/opc/blob/master/README.md)
they mention a OPCDAAuto.dll but I couldn't find where this dll is uptodate neither where can I download it.

**Reply by renzop** [reply]
Would be great if you can participate! We just use an OPC server from one particular vendor where we have controllers to test it. Have you seen the file [create_dll_interface.py](https://github.com/iterativ/openopc2/blob/develop/scripts/create_dll_interface.py) with this script you can create a python file that describes the interface of the dll that you find. As they write in the OPC Go library you can possibly use files from any OPC server to act as a wrapper, since the interface is standart.

**Reply by ldariva** [reply]
I facing this problem with a Yokogawa ExaOPC Server at the moment but I will face it in a lot of other vendor. The Yokogawa representative mentioned something about using a dll from ExaOPC server but he was not able to tell me exactly how. I will try to search for it in one of our OPC servers. 
One question. Where it the OpenOPC package we point to this dll? Sorry if the question looks silly. I am a chemical engineer and not a CS.
I never saw this python script but I will take a look on that.

**Reply by renzop** [reply]
;-) not a silly question. Actually there is no explicit "linking" in the code to that DLL. But in the file da_com.py on line 65:

`win32com.client.gencache.EnsureDispatch(opc_class, 0)`

This is where the com interface connection to that program (the OPC server or the wrapper dll is created). So by defining the a different opc_class you can communicate with other programs. (you should read about the COM interface first, otherwise it looks a bit too automagic to you). Then when you see parts like this:

` self.opc_client.CreateBrowser()`

The method CreateBrowser is actually a method exposed by that dll. And that method name and signature is defined by the OPC DA Specification (the document from the OPC foundation).

**Reply by ldariva** [reply]
Yes. After posting the last message I dig a little bit more and I saw that there other classe in this OPC_CLASS entry in windows register. I wonder why we can't use, for example, the Matrikon OPC Explorer class that is there (We use it to test connections and it is conneting to this OPC Server where we apply the DCOM patch. The only thing is that it is working just in the Syncronous mode). I am thinking about this because the Matrikon OPC Explorer must have implemented this CreateBrowser method in some way to work. I wonder if we could use it instead of the Gray-Box dll.

**Reply by renzop** [reply]
Yea give it a shot, just make sure that we can only use solutions here, were the licence for a particular software fits open source criterias.

**Reply by duduyoyo** [reply]
It is very hard to get support for dlls released decades ago (like ones from graybox or OPCDAAuto.dll). Vendors wouldn't spend any resource on them if they can't make profit (some vendors may be out of business). DCOM hardening is always a pain for COM-related application like classic OPC, especially if you don't have source code for any wrapper dlls. 

Check [this](https://github.com/duduyoyo/WebSocket4OPC) out. It provides a unique way to use Python to retrieve OPC DA data without any 3rd party wrapper dll. It is DCOM free so you wouldn't worry like this authentication level issue in future anymore. No gateway service installation either. Hope it can shed some lights on your issue!

**Reply by renzop** [reply]
Hi @duduyoyo  you are right about your first point. And I apreciate that you have your own open OPC Library. If you would like to contribute to OpenOPC2 and make it independent from the dll wrapper... Go on. I opened the discussion [Alternatives Open OPC2](https://github.com/iterativ/openopc2/discussions/53) So that people can promote their projects. (And don't need to put it in issues etc. :-) )

**Reply by duduyoyo** [reply]
Thanks for creating a new discussion. It is always beneficial when user has an alternative when one doesn't work. We have the same goal to help user get real value from unsupported legacy OPC in a modern way. You are welcome to promote OpenOPC2 in discussion of my repo in the way you want. Cheers!
